File: backend/app/core/roles.py
from __future__ import annotations

ADMIN_ROLE = "admin"
SOFTWARE_TEAM_ROLE = "software_team"
MANAGEMENT_ROLE = "management"
IT_ROLE = "it"
DRONE_ROLE = "drone"
FINANCE_ROLE = "finance"
HR_ROLE = "hr"
BD_ROLE = "bd"
ORTHO_ROLE = "ortho"
LIDAR_ROLE = "lidar"
CIVIL_ROLE = "civil"
LASER_SCANNING_ROLE = "laser_scanning"
BIM_ROLE = "bim"
MOBILE_MAPPING_ROLE = "mobile_mapping"
EMPLOYEE_ROLE = "employee"

VALID_ROLES = {
    ADMIN_ROLE,
    SOFTWARE_TEAM_ROLE,
    MANAGEMENT_ROLE,
    IT_ROLE,
    DRONE_ROLE,
    FINANCE_ROLE,
    HR_ROLE,
    BD_ROLE,
    ORTHO_ROLE,
    LIDAR_ROLE,
    CIVIL_ROLE,
    LASER_SCANNING_ROLE,
    BIM_ROLE,
    MOBILE_MAPPING_ROLE,
    EMPLOYEE_ROLE,
}

ADMIN_EQUIVALENT_ROLES = {ADMIN_ROLE, SOFTWARE_TEAM_ROLE}


def is_admin_equivalent(role: str | None) -> bool:
    return (role or "").strip().lower() in ADMIN_EQUIVALENT_ROLES


def role_is_allowed(user_role: str, allowed_roles: set[str]) -> bool:
    normalized = user_role.strip().lower()
    if normalized in allowed_roles:
        return True
    # Software Team keeps every permission that was previously granted to Admin.
    return normalized == SOFTWARE_TEAM_ROLE and ADMIN_ROLE in allowed_roles


def role_display_name(role: str) -> str:
    normalized = role.strip().lower()
    labels = {
        SOFTWARE_TEAM_ROLE: "Software Team",
        ADMIN_ROLE: "Admin",
        MANAGEMENT_ROLE: "Management",
        IT_ROLE: "IT Department",
        DRONE_ROLE: "Drone Department",
        FINANCE_ROLE: "Finance Department",
        HR_ROLE: "HR Department",
        BD_ROLE: "Business Development",
        ORTHO_ROLE: "Ortho",
        LIDAR_ROLE: "LiDAR",
        CIVIL_ROLE: "Civil Department",
        LASER_SCANNING_ROLE: "Laser Scanning",
        BIM_ROLE: "BIM Department",
        MOBILE_MAPPING_ROLE: "Mobile Mapping",
        EMPLOYEE_ROLE: "Employee Support",
    }
    return labels.get(normalized, normalized.replace("_", " ").title())


def backup_role(role: str) -> str:
    """Map full-access Software Team exports to the existing Admin export scope."""
    return ADMIN_ROLE if role.strip().lower() == SOFTWARE_TEAM_ROLE else role.strip().lower()

File: frontend/src/lib/roles.ts
import type { Role } from '../types'

export const FULL_ACCESS_ROLES: Role[] = ['software_team', 'admin']

export function isFullAccessRole(role: Role): boolean {
  return FULL_ACCESS_ROLES.includes(role)
}

// One shared operational workflow (BD -> Finance -> PM -> Team Lead -> Production/QC/QA ->
// Delivery -> Billing) covers all five technical Project Manager roles. Use this helper instead
// of scattering role === 'lidar' / 'civil' / ... checks across components.
export const TECHNICAL_PM_ROLES: Role[] = ['ortho', 'lidar', 'mobile_mapping', 'laser_scanning', 'civil']

export function isTechnicalProjectManager(role: Role): boolean {
  return TECHNICAL_PM_ROLES.includes(role)
}

// Finance Command Center: management/admin/finance see all departments; technical PMs see only their department.
export const COMMAND_CENTER_ROLES: Role[] = ['finance', 'admin', 'management', ...TECHNICAL_PM_ROLES]

const ROLE_DEPARTMENT_CODE: Partial<Record<Role, string>> = {
  ortho: 'ortho',
  lidar: 'lidar',
  mobile_mapping: 'mobile_mapping',
  laser_scanning: 'laser_scanning',
  civil: 'civil',
}

export function departmentCodeForRole(role: Role): string | null {
  return ROLE_DEPARTMENT_CODE[role] ?? null
}

export function canAccessRole(userRole: Role, allowedRoles?: Role[]): boolean {
  if (!allowedRoles || allowedRoles.length === 0) return true
  if (allowedRoles.includes(userRole)) return true
  return userRole === 'software_team' && allowedRoles.includes('admin')
}

export function roleHomePath(role: Role): string {
  const paths: Record<Role, string> = {
    software_team: '/software-team',
    admin: '/admin',
    management: '/management',
    it: '/it',
    drone: '/drone',
    finance: '/finance',
    hr: '/hr/travel-km',
    bd: '/bd',
    ortho: '/ortho',
    // Same shared operational dashboard as Ortho (one workflow, department-aware inside).
    lidar: '/ortho',
    civil: '/ortho',
    laser_scanning: '/ortho',
    mobile_mapping: '/ortho',
    bim: '/project-workstreams',
    employee: '/support',
  }
  return paths[role]
}

export function roleDisplayName(role: Role): string {
  const labels: Record<Role, string> = {
    software_team: 'Software Team',
    admin: 'Admin',
    management: 'Management',
    it: 'IT Department',
    drone: 'Drone Department',
    finance: 'Finance Department',
    hr: 'HR Department',
    bd: 'Business Development',
    ortho: 'Ortho',
    lidar: 'LiDAR',
    civil: 'Civil Department',
    laser_scanning: 'Laser Scanning',
    bim: 'BIM Department',
    mobile_mapping: 'Mobile Mapping',
    employee: 'Employee Support',
  }
  return labels[role]
}
